lib/path/default.nix

```diff
@@ -149,6 +173,54 @@ in /* No rec! Add dependencies on this file at the top. */ {
           ${subpathInvalidReason subpath}'';
     path + ("/" + subpath);
 
+  /*
+  Whether the second path is a prefix of the first path, or equal to it.
+  Throws an error if the paths don't share the same filesystem root.
+
+  Laws:
+
+  - Equivalent to whether some subpath exists that can be appended to the first path to get the second path:
+
+        hasPrefix p q <-> exists s . append p s == q
```

Comment:
Or double equal sign like the preceding law in the diff
```suggestion
        hasPrefix p q
          ==
        ∃s. (append p s == q)
```

I think the implementation could perhaps be optimized for the `true` case, because then you don't need to find the other root, or allocate path component lists, but you'd have to measure it. Something for a later PR maybe.

So just the suggestion (either one) and then lgtm

Reply:
I think I'd rather remove the extra newlines to be consistent with the rest of this file, sounds good otherwise!

Reply:
I'm curious what people think, so [unscientific twitter poll](https://twitter.com/robert_hensing/status/1669382584759926785).

Reply:
We could also do this to avoid the $\exists$:

> Equivalent to whether some subpath `s` exists that can be appended to the first path to get the second path:
>
>     hasPrefix p q == (append p s == q)

But also, the two `==`'s do look a bit weird, I think $\iff$ would be better, but yeah it's not that well known, I don't mind either way

Reply:
`<=>` is what I would expect.

Reply:
How about just having this, no additional formula:

> `hasPrefix p q` is only true if `q == append p s` for some subpath `s`.

Reply:
I think that's best, overall.

Though personally any of the notation above is fine too. I've finally stopped overriding `\iff` from `<=>` to `<->` in LaTeX :joy: